libuavcan/include/uavcan/introspection.hpp

```diff
@@ -0,0 +1,45 @@
+/*
+ * Optional utilites to provide introspection into libuavcan
+ * objects and to provide metrics for library performance.
+ * Facilities defined in this header can be compiled out of
+ * production code and should be used for debugging or targeted
+ * testing activities.
+ * 
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ */
+
+#ifndef UAVCAN_INTROSPECTION_HPP_INCLUDED
+#define UAVCAN_INTROSPECTION_HPP_INCLUDED
+
+#include "uavcan/uavcan.hpp"
+
+
+#ifndef UAVCAN_INTROSPECTION_TRACE_ENABLE
+# define UAVCAN_INTROSPECTION_TRACE_ENABLE 0
+#endif
+
+/*
+ * TODO: find a way to do this using SFINAE. Minimize C-preprocessor usage. 
```

Comment:
I doubt it's doable without the preprocessor, since SFINAE depends only on the supplied or deduced template arguments, meaning that the invoking code will have to select the appropriate version at the time of instantiation. The only possibility that comes to mind is to provide a macro the value of which alternates between traced and non-traced builds and use it in place of a template argument, but then again, how is it better than what you already have here -- we'd still have to rely on the preprocessor. I may be missing something though. It's been a while since I've done any serious C++-fu myself.

Reply:
Okay. I was just writing notes to my future self here anyway. Let's see what pans out when I get here.